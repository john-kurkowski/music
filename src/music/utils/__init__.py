"""Misc. utilities."""

import asyncio
import os
import shutil
from collections.abc import Callable, Iterator
from functools import wraps
from pathlib import Path
from typing import Any, NoReturn


def assert_exhaustiveness(no_return: NoReturn) -> NoReturn:  # pragma: no cover
    """Provide an assertion at type-check time that this function is never called."""
    raise AssertionError(f"Invalid value: {no_return!r}")


def coro(f: Callable[..., Any]) -> Callable[..., Any]:
    """Decorate Click commands as coroutines.

    H/T https://github.com/pallets/click/issues/85
    """

    @wraps(f)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        return asyncio.run(f(*args, **kwargs))

    return wrapper


def recurse_property[T](prop: str, obj: T | None) -> Iterator[T]:
    """Recursively yield the given optional, recursive property, starting with the given object."""
    while obj is not None:
        yield obj
        obj = getattr(obj, prop, None)


def rm_rf(path: Path) -> None:
    """Delete a file or directory recursively, if it exists, similarly to `rm -rf <PATH>`."""
    if os.path.isdir(path) and not os.path.islink(path):
        shutil.rmtree(path)
    elif os.path.exists(path):
        os.remove(path)
