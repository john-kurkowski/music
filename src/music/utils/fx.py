"""Utilities for working with Reaper's FX."""

import warnings

with warnings.catch_warnings():
    warnings.filterwarnings("ignore", message="Can't reach distant API")
    import reapy


def set_param_value(param: reapy.core.FXParam, value: float) -> None:
    """Set a parameter's value.

    Works around bug with reapy 0.10's setter.
    """
    parent_fx = param.parent_list.parent_fx
    parent = parent_fx.parent
    param.functions["SetParamNormalized"](  # type: ignore[operator]
        parent.id, parent_fx.index, param.index, value
    )
