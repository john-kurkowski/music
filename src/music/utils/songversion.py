"""Enum of different versions of a song to render."""

import enum
from pathlib import Path

from . import assert_exhaustiveness


class SongVersion(enum.Enum):
    """Different versions of a song to render."""

    MAIN = enum.auto()
    INSTRUMENTAL = enum.auto()
    INSTRUMENTAL_DJ = enum.auto()
    ACAPPELLA = enum.auto()
    STEMS = enum.auto()

    def name_for_project_dir(self, project_dir: Path) -> str:
        """Name of the project for the given song version."""
        project_name = project_dir.name
        if self is SongVersion.MAIN:
            return project_name
        elif self is SongVersion.INSTRUMENTAL:
            return f"{project_name} (Instrumental)"
        elif self is SongVersion.INSTRUMENTAL_DJ:
            return f"{project_name} (DJ Instrumental)"
        elif self is SongVersion.ACAPPELLA:
            return f"{project_name} (A Cappella)"
        elif self is SongVersion.STEMS:
            return f"{project_name} (Stems)"
        else:  # pragma: no cover
            assert_exhaustiveness(self)

    def path_for_project_dir(self, project_dir: Path) -> Path:
        """Path of the rendered file for the given song version."""
        path = project_dir / self.name_for_project_dir(project_dir)
        if self is SongVersion.STEMS:
            return path

        return Path(f"{path}.wav")

    @property
    def pattern(self) -> list[Path]:
        """Reaper directory render pattern for the given song version, if any."""
        if self is SongVersion.STEMS:
            # Roughly create a directory tree matching the tracks and folders in the Reaper project.
            return [Path("$folders $tracknumber - $track")]

        return []
