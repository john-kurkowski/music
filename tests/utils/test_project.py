"""Util tests."""

from pathlib import Path
from unittest import mock

import pytest
from syrupy.assertion import SnapshotAssertion

import music.utils.project

from ..conftest import RequestsMocks


@mock.patch("reapy.core.Project.__init__")
def test_project_reaper_not_running(core_project: mock.Mock) -> None:
    """Test handling when Reaper is not running."""
    core_project.side_effect = AttributeError("module doesn't have reascript_api, yo")
    with pytest.raises(Exception, match="Reaper running"):
        music.utils.project.ExtendedProject()
    with pytest.raises(Exception, match="Reaper running"):
        music.utils.project.ExtendedProject.get_or_open(Path("some path"))


@mock.patch("reapy.core.Project.__init__")
@mock.patch("reapy.RPR.EnumProjects", create=True)
def test_project_path(
    enum_projects: mock.Mock,
    core_project: mock.Mock,
    tmp_path: Path,
) -> None:
    """Test getting the path containing the project (not the _recording_ path)."""
    some_path = tmp_path / "path" / "to" / "some project"
    some_file = some_path / "some project.rpp"
    enum_projects.return_value = ("", "", some_path / some_file)
    obj = music.utils.project.ExtendedProject()
    assert obj.path == str(some_path)


@mock.patch("reapy.core.Project.__init__")
@mock.patch("reapy.RPR.EnumProjects", create=True)
@mock.patch("reapy.RPR.Main_openProject", create=True)
def test_project_open(
    open_project: mock.Mock,
    enum_projects: mock.Mock,
    core_project: mock.Mock,
    snapshot: SnapshotAssertion,
    tmp_path: Path,
) -> None:
    """Test invoking Reaper to open a project."""
    some_path = tmp_path / "path" / "to" / "some project"
    music.utils.project.ExtendedProject.get_or_open(some_path)
    assert open_project.mock_calls == snapshot


@pytest.mark.asyncio
@mock.patch("reapy.core.Project.__init__")
async def test_project_render(
    open_project: mock.Mock, requests_mocks: RequestsMocks, snapshot: SnapshotAssertion
) -> None:
    """Test invoking Reaper to render a project."""
    await music.utils.project.ExtendedProject().render()
    assert requests_mocks.mock_calls == snapshot
