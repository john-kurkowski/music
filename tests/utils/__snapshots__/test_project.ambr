# serializer version: 1
# name: test_project_open
  _CallList([
    _Call(
      '',
      tuple(
        'TMP_PATH_HERE/path/to/some project/some project.rpp',
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_project_render
  dict({
    'get': _CallList([
      _Call(
        '',
        tuple(
          'http://localhost:2307/_/42230',
        ),
        dict({
          'timeout': ClientTimeout(total=3600, connect=None, sock_read=None, sock_connect=None, ceil_threshold=5),
        }),
      ),
    ]),
    'post': _CallList([
    ]),
    'put': _CallList([
    ]),
  })
# ---
